dask/dataframe/multi.py

```diff
@@ -921,8 +926,24 @@ def stack_partitions(dfs, divisions, join="outer"):
     for df in dfs:
         # dtypes of all dfs need to be coherent
         # refer to https://github.com/dask/dask/issues/4685
+        # and https://github.com/dask/dask/issues/5968.
+        if is_dataframe_like(df):
+
+            shared_columns = df.columns.intersection(meta.columns)
+            needs_astype = [
+                col
+                for col in shared_columns
+                if df[col].dtype != meta[col].dtype
+                and not is_categorical_dtype(df[col].dtype)
+            ]
+
+            if needs_astype:
+                # Copy to avoid mutating the caller inplace
+                df = df.copy()
```

Comment:
is a deep copy necessary here ?  If not, can we using `df = df.copy(deep=False)`

Reply:
This is a dask DataFrame, so we aren't actually copying data here (and dask.dataframe doesn't have a `deep=True/False`). We just need to break the reference to the user's original DataFrame if we need to update a dtype.

Reply:
Cool.  Thanks @TomAugspurger